### PR TITLE
NMS-11979: Implement Default Focus

### PR DESCRIPTION
**Pull request by patrick-schweizer**
(separation of logic and data)

### All Contributors

* [X] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [X] Have you made a comment in that issue which points back to this PR?
* [X] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [ ] If this is a new or updated feature, is there documentation for the new behavior?
* [X] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

Once the reviewer(s) accept the PR and the branch passes continuous integration in Bamboo, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-11979
* Bamboo (Continuous Integration): https://bamboo.opennms.org/

### CODE REVIEW

**Comment by patrick-schweizer** [comment]
@mvrueden I took a first shot and did the following:
I separated the logic of determining the default focus (FocusStrategy) from the result (list of vertices). I believe this could give us the following advantages:
* easier to persist and transport (logic is hard to persist)
* flexibility in how the fdefault focus is determined (we are not necessarily bound to the [Focus](https://github.com/OpenNMS/opennms/pull/2557/files#diff-62c35d1acfc67180bab371dbdbad4adb) interface to determine a list of vertices)
* the downside is that the calculation is not done when getDefaultFocus() is called on a graph but normally when the graph is created. The difference is that the graph may have changed after the calculation and thus the focus list may not be the same as if it were calculated on the updated graph. But I assume that this is not a problem since the default focus calculation is only needed before a graph is displayed.

Before I continue (e.g. add persistence) please take a look and give me your input. Thanks!

File: features/graph/api/src/main/java/org/opennms/netmgt/graph/api/focus/FocusStrategy.java [hunk 1]
```diff
@@ -39,27 +39,47 @@
...
 public class FocusStrategy {
-    public static final Focus ALL = graphContext -> graphContext.getGraph().getVertices();
-
-    public static final Focus EMPTY = graphContext -> Lists.newArrayList();
+    
+    public static final Focus ALL = new Focus() {
+        
+        @Override
+        public <V extends Vertex> List<V> getFocus(Graph<V, ? extends Edge> graph) {
+            return graph.getVertices().parallelStream().collect(Collectors.toList());
```
Comment (mvrueden) [inline features/graph/api/src/main/java/org/opennms/netmgt/graph/api/focus/FocusStrategy.java]: Please don't use `parallelStream`.

**Comment by mvrueden** [comment]
This looks good to me in general.

We could let the Focus implement some kind of `Persistable` interface and delegate to the focus when persisting.

**Comment by mvrueden** [comment]
At the moment this is put on hold, until we have time to come back to this.

**Comment by mvrueden** [comment]
I reworked this so it works with the latest changes from `features/graph-service`. It also exposes the focus through ReST and persists it properly. 
@patrick-schweizer can you take a look?

File: features/graph/service/src/main/java/org/opennms/netmgt/graph/persistence/mapper/EntityToGenericMapper.java [hunk 2]
```diff
@@ -91,6 +92,16 @@ public GenericGraph fromEntity(final GraphEntity graphEntity) {
             genericGraphBuilder.addEdge(genericEdge.build());
         });
 
+        final FocusEntity focusEntity = graphEntity.getDefaultFocus();
+        if (focusEntity == null || focusEntity.getType().equalsIgnoreCase("EMPTY")) {
```
[the remark this answers is not in the document]
Comment (pbrane) [reply features/graph/service/src/main/java/org/opennms/netmgt/graph/persistence/mapper/EntityToGenericMapper.java]: EMPTY.isEmpty()

:)

**Comment by mvrueden** [comment]
Merging for now. If you have any comments @patrick-schweizer let me know and I will work it in in the next batch.

**Comment by SummaCumWilly** [comment]
So as a Topology UI user, how do I assign a node to be a "Default Node" for all users? I don't see where this is exposed in the Topology UI.

**Comment by dino2gnt** [comment]
@SummaCumWilly This isn't the correct venue for user questions. Please use our Discourse instead.  https://opennms.discourse.group/c/community-support/7

**Pull request merged**